deployment/terraform/agent.go

```diff
@@ -85,83 +86,113 @@ func (t *Terraform) configureAndRunAgents(extAgent *ssh.ExtAgent) error {
+				{srcData: strings.TrimPrefix(clientSysctlConfig, "\n"), dstPath: "/etc/sysctl.conf"},
+				{srcData: strings.TrimPrefix(limitsConfig, "\n"), dstPath: "/etc/security/limits.conf"},
+				{srcData: strings.TrimPrefix(prometheusNodeExporterConfig, "\n"), dstPath: "/etc/default/prometheus-node-exporter"},
 			}
 
-			// The new entry in /etc/hosts will make SiteURL point to:
-			// - The first instance's IP if there's a single node
-			// - The proxy's IP if there's more than one node
-			ip := output.Instances[0].PrivateIP
-			if output.HasProxy() {
-				ip = output.Proxy.PrivateIP
+			if t.config.UsersFilePath != "" {
+				batch = append(batch, uploadInfo{srcData: strings.Join(splitFiles[agentNumber], "\n"), dstPath: dstUsersFilePath, msg: "Uploading list of users credentials"})
 			}
 
-			proxyHost := fmt.Sprintf("%s %s\n", ip, t.config.SiteURL)
-			appHostsFile := fmt.Sprintf(appHosts, proxyHost)
+			// If SiteURL is set, update /etc/hosts to point to the correct IP
+			if t.config.SiteURL != "" {
+				output, err := t.Output()
+				if err != nil {
+					errChan <- err
+				}
+
+				// The new entry in /etc/hosts will make SiteURL point to:
+				// - The first instance's IP if there's a single node
+				// - The proxy's IP if there's more than one node
+				ip := output.Instances[0].PrivateIP
+				if output.HasProxy() {
+					ip = output.Proxy.PrivateIP
+				}
+
+				proxyHost := fmt.Sprintf("%s %s\n", ip, t.config.SiteURL)
+				appHostsFile := fmt.Sprintf(appHosts, proxyHost)
+
+				batch = append(batch, uploadInfo{srcData: appHostsFile, dstPath: "/etc/hosts", msg: "Updating /etc/hosts to point to the correct IP"})
+			}
 
-			batch = append(batch, uploadInfo{srcData: appHostsFile, dstPath: "/etc/hosts", msg: "Updating /etc/hosts to point to the correct IP"})
-		}
+			if err := uploadBatch(sshc, batch); err != nil {
+				errChan <- fmt.Errorf("batch upload failed: %w", err)
+			}
 
-		if err := uploadBatch(sshc, batch); err != nil {
-			return fmt.Errorf("batch upload failed: %w", err)
-		}
+			if out, err := sshc.RunCommand("sudo sysctl -p"); err != nil {
+				errChan <- fmt.Errorf("error running command, got output: %q: %w", out, err)
+			}
 
-		if out, err := sshc.RunCommand("sudo sysctl -p"); err != nil {
-			return fmt.Errorf("error running command, got output: %q: %w", out, err)
-		}
+			mlog.Info("Starting load-test api server", mlog.Int("agent", agentNumber))
+			if out, err := sshc.RunCommand("sudo systemctl daemon-reload && sudo service ltapi restart"); err != nil {
+				errChan <- fmt.Errorf("error running command, got output: %q: %w", out, err)
+			}
+		}()
+	}
 
-		mlog.Info("Starting load-test api server")
-		if out, err := sshc.RunCommand("sudo systemctl daemon-reload && sudo service ltapi restart"); err != nil {
-			return fmt.Errorf("error running command, got output: %q: %w", out, err)
-		}
+	// Run another goroutine to check when all agents have finished configuring
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
```

Comment:
I would restructure this a bit. Make `wg.Wait()` to be synchronous and remove the err chan. Closing the `errChan` while other goroutines are still sending data to it is actually a bug and will panic. 

What I'd suggest is:

```go
var wg sync.WaitGroup
for i, val := range agents {
  wg.Add(1)
  go func(i int, val string) {
    defer wg.Done()
    // do stuff here
    // log error as usual, but just set an atomic.Bool to true.
  }(i, val) // This is important to localize scope. Then you don't need the weird variable copying style.
}

wg.Wait()
if atomic.Bool == true, then return error. 
```